jvm/core/src/main/scala/org/scalatest/prop/RoseTree.scala

```diff
@@ -0,0 +1,208 @@
+/*
+ * Copyright 2001-2020 Artima, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.scalatest.prop
+
+import scala.annotation.tailrec
+import scala.concurrent.{Future, ExecutionContext}
+import org.scalactic.ColCompatHelper.LazyListOrStream
+
+trait RoseTree[+T] { thisRoseTreeOfT =>
+
+  val value: T
+
+  // Compute the shrinks list on demand using this RoseTree's value.
+  // This will be called only when a property fails, and just once, and it
+  // won't take long, so no need to make this a lazy val.
+  def shrinks: LazyListOrStream[RoseTree[T]]
+
+  private lazy val maximumIterationCount = 1000000
+
+  def depthFirstShrinks[E](fun: T => (Boolean, Option[E])): (LazyListOrStream[RoseTree[T]], Option[E]) = {
```

Comment:
Let's call this shrinkSearch, because it really is neither depth first nor breadth first.